### DISCUSSION TITLE
[RFC] ConfigProvider support `componentClassName`

### DISCUSSION THREAD

**Discussion by MadCcc** [ant-design/ant-design#39862] [RFCs - archive]
## Summary

ConfigProvider 新增 prop `componentClassName`，用于全局为组件添加自定义 `className`

## Motivation

组件定制化时往往会为组件添加一些自定义的 `className`。以往会推荐对 antd 的组件进行二次封装，但是这样做会对用户产生过多的心智负担，比如需要强制开发者从自定义组件引入等等，并且如果是简单的改造这样做就会显得冗余。因此我们在 ConfigProvider 上提供能够全局赋予组件 `className` 的属性，以方便自定义样式。

## API

**ConfigProvider**

| 参数 | 说明 | 类型 | 默认值 | 
| --- | --- | --- | --- | 
| button (example for components) | 组件全局配置 | `{ className?: string }` | - | 

## Basic Example

为所有 Button 组件添加 `custom-button` 的 `className`。

```tsx
import React from 'react';
import { Button, ConfigProvider } from 'antd';

export default () => (
  <ConfigProvider button={{ className: 'custom-button' }}>
    <Button>button</Button>
  </ConfigProvider>
)
```

## Detailed Design

所有在 ConfigProvider 上自定义的全局组件 `className` 都会由 context 传到对应组件，并添加到对应的 `className` 属性上。

组件逐个支持。

**Comment by zombieJ** [comment]
现在的 ConfigProvider 已经有 [Component] 前缀了，是不是直接复用：
```tsx
<ConfigProvider form={{ className: 'hello-world', autoComplete: 'off' }} />
```

**Reply by MadCcc** [reply]
那就是转成分组件支持了

**Reply by vagusX** [reply]
这样看着是设置全局默认选项？

**Comment by Wxh16144** [comment]
确实在二次封装有统一 `className` 的需求。 感觉添加一个和 `prefixCls` 功能类似的来满足。 比如叫 `extraPrefixCls`:

```tsx
import React from 'react';
import { Button, ConfigProvider } from 'antd';

export default () => (
  <ConfigProvider extraPrefixCls="company-space">
    <Button>button</Button>
  </ConfigProvider>
)
```

实际结果为自动拼接组件名 `${company-space}-${component-name}`

**Reply by MadCcc** [reply]
有 `prefixCls` 了

**Reply by Wxh16144** [reply]
> 有 `prefixCls` 了

这个 `prefixCls` 修改的是 ant 组件的 className， 在 4.x 如果我们改了 prefixCls ， 然后 less 里面变量也需要修改为相同的。

这里讨论的应该是我们期望一个约定的不带任何样式的 className，可以这么理解 `extraPrefixCls` 只是针对二次封装的 className 前缀

**Reply by MadCcc** [reply]
这个方案的侵入性很高，相当于所有用过 prefixCls 的 classname 都要有一份 copy 出来，有些重了

**Reply by Wxh16144** [reply]
这里应该仅仅只是修改根组件 prefixCls, 不应该等于 copy。 说起来有点类似 namespace ？

**Comment by JarvisArt** [comment]
这个 componentClassName 是指组件级别的 prefixCls，还是就是className？如果是className，我指定了是不是默认的样式就都失效了

**Comment by JarvisArt** [comment]
我知道了，是全局为 Button 组件添加额外的 className

**Comment by crazyair** [comment]
要不要直接支持 modify component props

**Reply by crazyair** [reply]
我需要设置 Modal size middle，其他组建全局设置 size small

**Reply by Wxh16144** [reply]
> 其他组建全局设置 size small

但是有些组件只有 `default` 和 `small`。 有些又是 `large` | `middle` | `small`。感觉也不好做

**Comment by afc163** [comment]
Activation => Motivation

**Comment by phatify** [comment]
+1

**Comment by amir2mi** [comment]
This approach can be extended to change components' default icons in one place, [as mentioned here](https://github.com/ant-design/ant-design/issues/51269)
[Related issue](https://github.com/ant-design/ant-design/issues/41413)
